design/kanister-kopia-integration/kanister-kopia-integration.md

```diff
@@ -0,0 +1,91 @@
+It is a lock-free system that allows concurrent multi-client operations including garbage collection.
+
+To explore other features of Kopia, see its [documentation](https://kopia.io/docs/features/).
+
+## Scope
+
+1. Automate the initialization of a Kopia Repository for an application backed up by Kanister.
+2. Design and automate the lifecycle of the required Kopia Repository Server.
+3. Add new versions of Kanister Data Functions like BackupData, RestoreData, etc. with Kopia as the primary data mover tool.
+
+## User Experience
+
+- All the new features mentioned in this document will be opt-in only. Existing users will not see any changes in the Kanister controller's behavior.
+- Users will be able to continue using their current Blueprints, switch to the v2 version of the example Blueprints, or use Blueprints with the new version of the Kanister Data Functions.
+- Users opting to use the v2 Blueprints and Blueprints with Kopia-based Kanister Data Functions will be required to follow instructions to set up the required Kopia Repository Server before executing the actions.
+- After setting up the Repository Server, users can follow the normal workflow to execute actions from the v2 Blueprints. To use the new versions of the Kanister Data Functions, users must specify the version of the function via the ActionSet Action's `preferredVersion` field.
+
+## Detailed Design
+
+### Kopia Repository
+
+- As mentioned above, the backup storage location is called a "Repository" in Kopia.
+- A separate repository will be used for each application protected by Kopia-based Blueprints in Kanister.
+- The repository will be initialized when the Repository Server is created the first time. Once initialized, it will continue to exist at the location until cleaned up manually.
+- Accessing the repository requires location and credential information similar to a Kanister Profile CR and a unique password used by Kopia during [encryption](https://kopia.io/docs/features/#end-to-end-zero-knowledge-encryption).
+- In the first iteration, this password will be auto-generated by the Kanister controller. Future iterations will allow users to use a Key Management Service of choice.
+- Only a single repository can exist at a particular backend storage location. To address this, the Kanister controller will generate the repository path using the location information and the UUID of the application namespace. For example, the repository path for `mysql` namespace and S3 bucket called `test-bucket` will be of the form `test-bucket/<UUID of mysql namespace>`.
+
+### Kopia Repository Server
+
+- A Kopia Repository Server allows Kopia clients proxy access to the backend storage location through it.
+- A separate server will be used for each repository initialized in Kanister.
```

Comment:
a separate server would be used just for initialising the repo or for every action?

Reply:
There will be a long-running server for each repository and not for every action. Let me know if you want me to modify the statement to clarify this.

Reply:
I think it's ok.